examples/platforms/simulation/radio.c

```diff
@@ -1321,3 +960,623 @@ otError otPlatRadioGetRegion(otInstance *aInstance, uint16_t *aRegionCode)
+    if (otPlatDiagModeGet())
+    {
+        otPlatDiagRadioTransmitDone(aInstance, frame, error);
+    }
+    else
+#endif
+    {
+        otPlatRadioTxDone(aInstance, frame, ack, error);
+    }
+}
+
+static void radioReceive(otInstance *aInstance)
+{
+    bool isTxDone = false;
+    bool isAck    = otMacFrameIsAck(&sReceiveFrame);
+
+    otEXPECT(sReceiveFrame.mChannel == sReceiveMessage.mChannel);
+    otEXPECT(sState == OT_RADIO_STATE_RECEIVE || sState == OT_RADIO_STATE_TRANSMIT);
+
+    // Unable to simulate SFD, so use the rx done timestamp instead.
+    sReceiveFrame.mInfo.mRxInfo.mTimestamp = otPlatTimeGet();
+
+    if (sTxWait && otMacFrameIsAckRequested(&sTransmitFrame))
+    {
+        isTxDone = isAck && otMacFrameGetSequence(&sReceiveFrame) == otMacFrameGetSequence(&sTransmitFrame);
+    }
+
+    if (isTxDone)
+    {
+        platformTransmitReturn(aInstance, &sTransmitFrame, (isAck ? &sReceiveFrame : NULL), OT_ERROR_NONE);
+    }
+    else if (!isAck || sPromiscuous)
+    {
+        radioProcessFrame(aInstance);
+    }
+
+exit:
+    return;
+}
+
+void radioSendMessage(otInstance *aInstance)
+{
+#if OPENTHREAD_CONFIG_MAC_HEADER_IE_SUPPORT && OPENTHREAD_CONFIG_TIME_SYNC_ENABLE
+    if (sTransmitFrame.mInfo.mTxInfo.mIeInfo->mTimeIeOffset != 0)
+    {
+        uint8_t *timeIe = sTransmitFrame.mPsdu + sTransmitFrame.mInfo.mTxInfo.mIeInfo->mTimeIeOffset;
+        uint64_t time = (uint64_t)((int64_t)otPlatTimeGet() + sTransmitFrame.mInfo.mTxInfo.mIeInfo->mNetworkTimeOffset);
+
+        *timeIe = sTransmitFrame.mInfo.mTxInfo.mIeInfo->mTimeSyncSeq;
+
+        *(++timeIe) = (uint8_t)(time & 0xff);
+        for (uint8_t i = 1; i < sizeof(uint64_t); i++)
+        {
+            time        = time >> 8;
+            *(++timeIe) = (uint8_t)(time & 0xff);
+        }
+    }
+#endif // OPENTHREAD_CONFIG_MAC_HEADER_IE_SUPPORT && OPENTHREAD_CONFIG_TIME_SYNC_ENABLE
+
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
+    if (sCslPeriod > 0 && !sTransmitFrame.mInfo.mTxInfo.mIsHeaderUpdated)
+    {
+        otMacFrameSetCslIe(&sTransmitFrame, (uint16_t)sCslPeriod, getCslPhase());
+    }
+#endif
+
+    sTransmitMessage.mChannel = sTransmitFrame.mChannel;
+
+    otEXPECT(radioProcessTransmitSecurity(&sTransmitFrame) == OT_ERROR_NONE);
+    otPlatRadioTxStarted(aInstance, &sTransmitFrame);
+    radioTransmit(&sTransmitMessage, &sTransmitFrame);
+
+    // Wait for simulator confirmation signal of radio transmission in virtual time mode.
+    sTxWait = true;
+exit:
+    return;
+}
+
+void setupTransmission(otInstance *aInstance)
+{
+    otError  error = OT_ERROR_NONE;
+    uint32_t delay;
+
+    otEXPECT_ACTION(!sRadioTransmitting, error = OT_ERROR_ALREADY);
+
+    radioSendMessage(aInstance);
+
+exit:
+    if (error == OT_ERROR_ALREADY)
+    {
+        // Radio is already transmiting, wait for it to end
+        if (sTransmittingUntil > otPlatTimeGet())
+        {
+            delay = (uint32_t)(sTransmittingUntil - otPlatTimeGet()) + 1;
```

Comment:
Why do we need a timer set for the transmission to end? If it ends in OTNS, an event 'tx done' will be generated and sent to the OT node at the right time. That will automatically wake up the node again and let it do event-message processing. 

Also, would it be safe to use a timer from the OT platform for this purpose? (E.g. is it separate from any other timers so that it doesn't interfere with other timed processes?)

Reply:
This is part of an error where the protocol does not wait for its state to change and get able to transmit something else.

The point of this code tries to mimic the exactly same behavior of a radio hardware code ([like this cc2538](https://github.com/openthread/ot-cc2538/blob/f8c16aa5fdf0490a84685b8b7d126838da42692b/src/radio.c#L505)), which would blocks the process with a ```while/until``` looking at a radio's register to be cleared. Thus, here, the device will sleep waiting for the remaining transmission time (which is actually updated by an incoming ```Tx Done``` signal from the OT-NS).

Reply:
About the 'safety' remark I made: I just looked into the OT code how timers are implemented, and it seems to be just adding one more entry to a linkedlist so shouldn't interfere with any other timers set by other OT processes. 

But still not sure why we would need timer code here: suppose you would NOT set the timer using otPlatAlarmMicroStartAt() - then still, the OTNS would send a TxDone event at the time that the transmission is done and the radio is possibly 'free' again.  Because the TxDone event is received, automatically the OT node events/process loop gets a chance to execute any pending processes or packet transmissions.  At exactly the right time. So the effect of that is equivalent to setting yourself a timer to wake up again at that time. But it doesn't hurt to have it in any case.

Reply:
The idea is to mimic a real radio hardware, blocking the code here. In fact, if this is not blocked here, some other parameters will behave differently. When we do not use this halt (going to sleep until), the radio returns an ERROR. The higher layer acts upon it, with a backoff or another series of attempts. This is not the real behavior.

I did not start this code with this timer. But, as I was developing, I soon realized this fault of not waiting at this point (as other radios would). That is why this stop is necessary, even if OTNS can correctly tell the device when the transmission is done.

Simulated devices are executed as another thread, and if no wait is implemented, they will skip this setup function right away, and do not wait for any OTNS msg.

Reply:
Ok thanks for the clarification. So the idea is to block any other tasks (radio, non-radio, all) from executing while the Tx is ongoing. This seems difficult to achieve, because other processes might register timers for an earlier time, and so these other processes may still get executed earlier than the time this timer is set for.

The otPlatAlarmMicroStartAt() only adds a time to the timer, leaving other times that were already set in place. When all tasks that were pending have run, control is passed back to OTNS in platform-sim.c by a blocking wait on the UDP socket which is the real moment the process blocks. To really simulate the radio Tx blocking every other task, there would need to be some change in platform-sim.c and/or the CLI app to suspend task processing in the main loop while Tx is still ongoing.  I also wonder if other radios than CC2538 do the same? Because it's a rather 'nasty' way of breaking the OT task system - I can check maybe at some point with OT authors about the intended design here.

Reply:
Correction on my previous comment: the otPlatAlarmMicroStartAt() does not add a time to the timer/scheduler, but I see now that it forces the next sleep/alarm event to be triggered at the specified time.  Simon had a comment before ([here](https://github.com/openthread/openthread/pull/7500#discussion_r950271013)) that only the Timer module is supposed to call this function. Basically it's the Timer::Scheduler::SetAlarm() method that sets the timer alarm of the simulation platform timer to the time of the first-to-occur timer event  in a list of events that it maintains. So the single 'alarm' value ought to be under control of the Timer module. Multiple processes may register an event with the Timer, and the Timer sorts these values so that a wake-up of the node can occur at the time of the earliest of these events. For this reason forcing the alarm value to some value seems potentially risky because scheduled tasks could be running late.

On the other hand this gives exactly the behavior of a blocking radio driver (that also forces some tasks to be run potentially late.) that postpones all remaining OT tasks until after the time that the Tx is done.

Reply:
Wow, Nice observation!

I saw that @simonlingoogle complained about that, but I needed to stop the process here to emulate the radio strategy. Your point on how this module impacts the code is very interesting.

I think that we need some sort of halt for the simulation here (can be my fault that I didn't see it). Either we define a safe method at the Timer module, or we create another approach (I don't have a suggestion now), but just OT-NS signals might now work.

Maybe I didn't see it, but I don't recall any comment recommending not to use this time as I did. I do agree with its potential risks, but nothing was clear back then when I implemented.

Thanks.

Reply:
@Vinggui I think your approach with the timer can just work, it will postpone any scheduled tasks until after the radio is done. I had an exchange with OT team members and the summary is that not all radio platforms will implement blocking waits, like the CC2538 code. Some implement with an RTOS where the low-level radio driver has its own thread. If the only app running on the node is the CLI (like in our case typically), then no accurate timing is needed for OT tasks for sure and 'blocking' is ok. The nice thing is that we can simulate nodes of both types - blocking or non-blocking - just by tweaking the code a bit.

When the CSL feature would be enabled (not supported by CC2538), accurate timing does become important. But still in this case the approach with the timer may just work, as CSL never schedules tasks in parallel with a transmission.